### DISCUSSION TITLE
Possible bug with opacity_unit_distance in add_volume()

### DISCUSSION THREAD

**Discussion by georg-walther** [pyvista/pyvista#3509] [Q&A]
Dear PyVista-Team,
Dear PyVista-Community,

When trying to add a volume with NaNs, there seems to be no straight-forward way to set the opacity for NaNs to 0 without affecting the remaining pixels.
To work around this problem I tried to set the parameter "opacity_unit_distance" to 0 so that the opacity is at its maximum immediately. As expected the opacity for the NaNs is 0 (see screenshot [1]) and maximal for the rest. But this is only the case if small values are far away from 0 in the original image. When those pixel values are close to 0 the opacity is also close to 0. Even adding 100 to all pixels and setting clim=[0, max(grid['scalars'])] does not work, which looks highly counterintuitive to me (see code example and screenshot [2]).

Any workaround to make NaNs invisble without affecting the rest is highly appreciated.
Many thanks in advance!
 
Code to reproduce possible bug:
```
import pyvista as pv
import numpy as np

noise = pv.perlin_noise(0.1, (1, 1, 1), (0, 0, 0))
grid = pv.sample_function(noise, [0, 1.0, -0, 1.0, 0, 1.0], dim=(20, 20, 20))

minV = min(grid['scalars'])
maxV = max(grid['scalars'])

grid['scalars'] = (grid['scalars'] - minV) / (maxV - minV)
grid['scalars'] *= 100

grid['scalars'] = np.where(grid['scalars'] < 20, np.nan, grid['scalars'])
grid['scalars'] = np.where(grid['scalars'] < 50, 1, grid['scalars'])

grid['scalars'] = grid['scalars'] + 100

p = pv.Plotter()
p.add_volume(grid, scalars='scalars', cmap="jet", clim=[0, max(grid['scalars'])], opacity_unit_distance=0)
p.show(window_size=[500, 500])
```

[1] How the image should look like:
![should look like this](https://user-images.githubusercontent.com/62775496/197585652-2097ee2d-12d7-4d04-9d16-64c94d479f7b.png)

[2] How it actually looks when some values are close to 0 but not 0:
![bug image](https://user-images.githubusercontent.com/62775496/197585880-7f53e544-f24e-4e01-bc63-678ef205cce4.png)

**Comment by georg-walther** [comment]
I can't use add_mesh() as a workaround because the 3D-image in my actual project has filigreed structures which are cut away when using add_mesh(). I assume this a because of marching cubes and a suboptimal isovalue for my 3D-image.

**Reply by banesullivan** [reply]
>has filigreed structures which are cut away when using add_mesh().

I'm not sure what this means, would you be able to share/demonstrate this issue?

**Reply by georg-walther** [reply]
To generate screenshot [3] I used your suggestion in ["Opacity for NaN-values when adding mesh"](https://github.com/pyvista/pyvista/discussions/3498) which is a threshold filter after converting to cell data. The mesh is sometimes broken, especially at the small vessels. Nevertheless, I think I was wrong about "marching cubes" and "isovalues". The conversion to cell data and the subsequent thresholding removes not only NaNs (background in my case) but also some vessel pixels.
Screenshot [4] was generated using `p.add_volume(img, scalars='scalars', cmap="jet", opacity='geom', opacity_unit_distance=0, show_scalar_bar=False) `and the small vessels are not broken/interrupted. The scalar array is binary. When it contains more values (e.g. [0,1,2, ..., 255]) the main issue of this discussion occurs.

Screenshot [3]:
![intrapulVesselMask-2D](https://user-images.githubusercontent.com/62775496/198847652-db3d4bc7-2843-4a7f-8648-759e8e966e85.jpeg)
Screenshot [4]:
![TESTintrapulVesselMask-2D](https://user-images.githubusercontent.com/62775496/198847654-3d0cc259-02c1-490b-a096-7d41f3fbeb96.jpeg)

**Comment by banesullivan** [comment]
Hm. this is tough. I'm seeing a very similar issue in ParaView too:

<img width="1602" alt="Screen Shot 2022-10-27 at 9 50 07 PM" src="https://user-images.githubusercontent.com/22067021/198448792-7f8eada3-769f-4fdf-ab3a-6f0e210d04a6.png">

**Comment by banesullivan** [comment]
Perhaps play with the `opacity` transfer function as the default is a linear map between 0 and 1. For example:

```py
p = pv.Plotter(notebook=0)
p.add_volume(grid, scalars='scalars', 
             opacity=[0.2, 0.5, 1.0],
             cmap="jet", clim=[0, max(grid['scalars'])], opacity_unit_distance=0.1)
p.show(window_size=[500, 500])
```
<img width="362" alt="Screen Shot 2022-10-27 at 9 55 25 PM" src="https://user-images.githubusercontent.com/22067021/198451395-4779a51d-9237-4508-b1fa-80006c7a3ab2.png">



but that might still be showing the NaN values.... 😕
